lib/adal-aad-node/src/engine/AzureAuth.ts

```diff
@@ -0,0 +1,469 @@
+		protected readonly stringLookup: StringLookup,
+		protected readonly azureAuthType: AzureAuthType
+	) {
+		this.clientId = providerSettings.id;
+		this.loginEndpointUrl = providerSettings.loginEndpoint;
+	}
+
+	protected abstract async login(tenant: Tenant, resource: AADResource): Promise<LoginResponse>;
+
+	public async startLogin(): Promise<AzureAccount | undefined> {
+		let loginComplete: Deferred<void> | undefined;
+		try {
+			const result = await this.login(this.commonTenant, this.providerSettings.resources.windowsManagementResource);
+			loginComplete = result?.authComplete;
+			if (!result?.response) {
+				this.logger.error('Authentication failed');
+				return undefined;
+			}
+			const account = await this.hydrateAccount(result.response.accessToken, result.response.tokenClaims);
+			loginComplete?.resolve();
+			return account;
+		} catch (ex) {
+			if (ex instanceof AzureAuthError) {
+				loginComplete?.reject(ex.getPrintableString());
+				// Let the caller deal with the error too.
+				throw ex;
+			}
+			this.logger.error(ex);
+			return undefined;
+		}
+	}
+
+	private getHomeTenant(account: AzureAccount): Tenant {
+		// Home is defined by the API
+		// Lets pick the home tenant - and fall back to commonTenant if they don't exist
+		return account.properties.tenants.find(t => t.tenantCategory === 'Home') ?? account.properties.tenants[0] ?? this.commonTenant;
+	}
+
+	public async refreshAccess(account: AzureAccount): Promise<AzureAccount> {
+		// Deprecated account - delete it.
+		if (account.key.accountVersion !== AzureAuth.ACCOUNT_VERSION) {
+			account.delete = true;
+			return account;
+		}
+		try {
+			const tenant = this.getHomeTenant(account);
+			const tokenResult = await this.getAccountSecurityToken(account, tenant.id, this.providerSettings.resources.windowsManagementResource);
+			if (!tokenResult) {
+				account.isStale = true;
+				return account;
+			}
+
+			const tokenClaims = this.getTokenClaims(tokenResult.token);
+			if (!tokenClaims) {
+				account.isStale = true;
+				return account;
+			}
+			return await this.hydrateAccount(tokenResult, tokenClaims);
+		} catch (ex) {
+			account.isStale = true;
+			return account;
+			// Let caller deal with it too.
+			throw ex;
+		}
+	}
+
+	public async hydrateAccount(token: Token | AccessToken, tokenClaims: TokenClaims): Promise<AzureAccount> {
+		const tenants = await this.getTenants({ ...token });
+		const account = this.createAccount(tokenClaims, token.key, tenants);
+		return account;
+	}
+
+	public async getAccountSecurityToken(account: AzureAccount, tenantId: string, azureResource: AADResource): Promise<Token | undefined> {
+		if (account.isStale === true) {
+			this.logger.log('Account was stale. No tokens being fetched.');
+			return undefined;
+		}
+		const tenant = account.properties.tenants.find(t => t.id === tenantId);
+
+		if (!tenant) {
+			throw new AzureAuthError(1, this.errorLookup.getTenantNotFoundError({ tenantId }), undefined);
```

Comment:
What's the 1 here? I don't see an error code with 1. We shouldn't be using magic numbers like this regardless - the enums make it a lot easier to work with